feat(api/v3): add GetDependencies RPC

Sapthagiri777 opened the pull request:
Part of jaegertracing/jaeger#7595

## Which problem is this PR solving?
Adds `GetDependencies` RPC to API v3 `query_service.proto` to enable service dependency graph queries in Jaeger v3 API. Required for jaegertracing/jaeger#7595.

## Description of the changes
- Added `GetDependencies(GetDependenciesRequest) returns (DependenciesResponse)` RPC to `proto/api_v3/query_service.proto`
- Added required proto messages: `GetDependenciesRequest`, `DependenciesResponse`, `Dependency`
- Matches v2 API structure: `end_time_nanos`, `lookback_nanos`, parent/child/call_count
- Generated code requires OpenTelemetry proto submodule (CI will handle)

## How was this change tested?
- Manual proto edit verified syntax
- `make proto` (v2 succeeded, v3 needs CI OpenTelemetry submodule)
- Ready for Jaeger backend implementation using new proto

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [ ] I have added unit tests for the new functionality *(proto change only)*
- [ ] I have run lint and test steps successfully *(CI will validate)*

Sapthagiri777 committed:
feat(api/v3): add GetDependencies RPC

Fixes jaegertracing/jaeger#7595

Signed-off-by: Sapthagiri777 <sriramsapthagiri@gmail.com>

Sapthagiri777 committed:
refactor: use google.protobuf.Timestamp for GetDependenciesRequest

Signed-off-by: Sapthagiri777 <sriramsapthagiri@gmail.com>

yurishkuro commented on proto/api_v3/query_service.proto:
```diff
@@ -136,6 +136,26 @@ message GetOperationsResponse {
+  google.protobuf.Timestamp start_time = 2;
+
+  // Optional. The end time to search trace ID.
+  google.protobuf.Timestamp end_time = 3;
```
```suggestion
  google.protobuf.Timestamp end_time = 2;
```

yurishkuro committed:
Apply suggestions from code review

Signed-off-by: Yuri Shkuro <yurishkuro@users.noreply.github.com>

Sapthagiri777 commented on proto/api_v3/query_service.proto:
```diff
@@ -136,6 +136,26 @@ message GetOperationsResponse {
 }
 
+message GetDependenciesRequest {
+    // Optional. The start time to search trace ID.
```
i checked the v2 code and it looks like both start_time and end_time are required — it throws an error if either one is missing.
Should I do the same for v3? Or would you like to set some defaults instead?
Let me know and I'll update the comments.

Sapthagiri777 replied:
Hi @yurishkuro! All done! 
Quick rundown of what I tweaked:
Switched both start_time and end_time to required — no more optional stuff
Updated the wording to say "search dependencies" (was saying "search trace ID" before, my bad!)
Bumped the field numbers to 1 and 2 to keep things clean
Here's how it looks now:
message GetDependenciesRequest {
  // Required. The start time for the time range to search dependencies.
  google.protobuf.Timestamp start_time = 1;
  // Required. The end time for the time range to search dependencies.
  google.protobuf.Timestamp end_time = 2;
}
Basically mirrors how v2 handles it — if someone skips either field, they'll get an error back.

Sapthagiri777 committed:
docs: mark start_time and end_time as required fields

Signed-off-by: Sapthagiri777 <sriramsapthagiri@gmail.com>

Sapthagiri777 committed:
fix: remove extra closing brace in proto file

Signed-off-by: Sapthagiri777 <sriramsapthagiri@gmail.com>

Sapthagiri777 commented:
Hey @yurishkuro! I noticed the test-code-gen check is failing, so I tried running make proto-api-v3-all locally to regenerate the files. But I ran into some errors with missing directories.
I'm still learning my way around the project — would you mind pointing me in the right direction? I'd really appreciate any guidance on how to fix this!

yurishkuro commented:
What errors?

The CONTRIBUTING.md should have the instructions.

Sapthagiri777 committed:
chore: regenerate swagger file

Signed-off-by: Sapthagiri777 <sriramsapthagiri@gmail.com>

Sapthagiri777 commented:
@yurishkuro! Thanks for pointing me to the CONTRIBUTING.md — that helped a lot! I ran make proto-all and it regenerated the swagger file. Just pushed the update.
Waiting for the CI checks to run now.

Sapthagiri777 committed:
Merge branch 'main' into feature/7595-api-v3-dependencies

Signed-off-by: SRIRAM SAPTHAGIRI <sriramsapthagiri@gmail.com>

Sapthagiri777 commented:
I've resolved the merge conflicts and committed the changes. Let me know if anything else needs fixing

yurishkuro commented on proto/api_v3/query_service.proto:
```diff
@@ -144,6 +144,26 @@ message GetOperationsResponse {
+
+
+message DependenciesResponse {
+  repeated Dependency dependencies = 1;
```
add REQUIRED annotation

Sapthagiri777 replied:
Done! Updated to latest main and added the REQUIRED annotation.

yurishkuro commented on proto/api_v3/query_service.proto:
```diff
@@ -144,6 +144,26 @@ message GetOperationsResponse {
+message Dependency {
+  string parent = 1;
+  string child = 2;
+  uint64 call_count = 3;
```
add REQUIRED annotations to all 3 fields

Sapthagiri777 replied:
Added REQUIRED annotations to all 3 fields

yurishkuro commented on proto/api_v3/query_service.proto:
```diff
@@ -187,11 +207,12 @@ service QueryService {
       get: "/api/v3/operations"
     };
   }
+   rpc GetDependencies(GetDependenciesRequest) returns (DependenciesResponse);
```
what's with the odd indentation?

Sapthagiri777 replied:
Sorry about that! Fixed the indentation now

yurishkuro reviewed (changes requested):
update to latest main and regenerate `make proto-all`

Sapthagiri777 committed:
fix: add REQUIRED annotations and fix indentation

Signed-off-by: Sapthagiri777 <sriramsapthagiri@gmail.com>

Sapthagiri777 commented:
@yurishkuro  , I've added the REQUIRED annotations and fixed the indentation.

Sapthagiri777 commented:
Hey @yurishkuro! I've addressed all the feedback — added REQUIRED annotations and fixed the indentation. Ready for another look when you get a chance! .

Sapthagiri777 commented:
Hi @yurishkuro, I've merged with latest main and regenerated with `make proto-all`. Everything is up to date now. Could you please approve the workflows and take another look? Thanks!

yurishkuro commented on proto/api_v3/query_service.proto:
```diff
@@ -187,11 +207,12 @@ service QueryService {
 
   // GetOperations returns operation names.
-  rpc GetOperations(GetOperationsRequest) returns (GetOperationsResponse) {
+   rpc GetOperations(GetOperationsRequest) returns (GetOperationsResponse) {
```
```suggestion
  rpc GetOperations(GetOperationsRequest) returns (GetOperationsResponse) {
```

yurishkuro committed:
Apply suggestion from @yurishkuro

Signed-off-by: Yuri Shkuro <yurishkuro@users.noreply.github.com>

yurishkuro commented on proto/api_v3/query_service.proto:
```diff
@@ -192,6 +212,7 @@ service QueryService {
       get: "/api/v3/operations"
     };
   }
+  rpc GetDependencies(GetDependenciesRequest) returns (DependenciesResponse);
```
missing http settings

Sapthagiri777 committed:
feat: add HTTP settings for GetDependencies RPC

Signed-off-by: Sapthagiri777 <sriramsapthagiri@gmail.com>

Sapthagiri777 commented:
Added the HTTP settings for GetDependencies! Let me know if anything else needs to be fixed.

yurishkuro commented:
Do not say "Fixes ..." if it's a partial PR, as it will close the issue.
